components/script/dom/node.rs

```diff
@@ -1416,7 +1536,13 @@ impl TreeIterator {
     }
 
     fn next_skipping_children_impl(&mut self, current: DomRoot<Node>) -> Option<DomRoot<Node>> {
-        for ancestor in current.inclusive_ancestors() {
+        let iter = if self.shadow_including {
+            current.shadow_including_inclusive_ancestors()
```

Comment:
Ah, I guess this is why you need the Box<>. Would be nice to either box here, or not box at all and make both iterators have the same type. Again can be followup.